### ISSUE TITLE
Unable to run terraform apply against a specific target vsphere_virtual_machine: Error: Provider produced inconsistent final plan

### ISSUE THREAD

**Issue by andrew-lee-metaswitch** [hashicorp/terraform-provider-vsphere#1125]
### Terraform Version
```
Terraform v0.12.26
+ provider.local v1.4.0
+ provider.vsphere v1.17.4
```
Also tested on (provider.vsphere 1.21.0)

### Affected Resource(s)

We have a use-case where in order to set up an etcd cluster correctly we must first spin up the first virtual machine in a `vsphere-virtual_machine` of count 3, then spin up the other VMs.

We want to do this my trying `terraform apply -target=module.<module_name>.vsphere_virtual_machine.vms[0]` then running `terraform apply` a second time. (As an aside: we can get round this by splitting off the "first VM" into a separate module resource but we'd rather not do that: it is useful for later lifecycle events to keep them as a single resource) 

However what we are seeing is that the first time we run ` `terraform apply -target=module.<module_name>.vsphere-virtual_machine.vms[0]` we get the following error, (although re-running this command works fine the second time:
```
Error: Provider produced inconsistent final plan

When expanding the plan for
module.mymodule.vsphere_virtual_machine.my_vms[0] to
include new values learned so far during apply, provider
"registry.terraform.io/-/vsphere" produced an invalid new value for
.disk[0].datastore_id: was cty.StringVal("<computed>"), but now
cty.StringVal("datastore-13037").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

Error: Provider produced inconsistent final plan

When expanding the plan for
module.mymoduke.vsphere_virtual_machine.myvmss[0] to
include new values learned so far during apply, provider
"registry.terraform.io/-/vsphere" produced an invalid new value for
.disk[0].thin_provisioned: was cty.False, but now cty.True.
```

This error looks similar to https://github.com/hashicorp/terraform-provider-vsphere/issues/1080 insomuch as it's hitting https://www.terraform.io/docs/extend/terraform-0.12-compatibility.html#inaccurate-plans

It looks heavily related to https://github.com/hashicorp/terraform-provider-vsphere/issues/1028  and 
https://github.com/hashicorp/terraform-provider-vsphere/pull/1075

I've tested with @jcpowermac's branch https://github.com/jcpowermac/terraform-provider-vsphere/tree/fixes_1028 and verified that that fixes my issue, although I note that essentially turns of all post-disk clone validation. 

So we have two problems with cloning disks:

1. Datastore ID changes
2. thin_provisioned changes,

In my example: the datastore is not changing, both the template and desired VM are thin provisioned.  I think about thin provisioning and eagerly scrub are two very related settings, so i think the second issue is a dupe of 1028.

I think we want to be able to say that when cloning a disk we don't really care if the datastore it lives on changes, right? 

### Terraform Configuration Files

```hcl
provider "vsphere" {
  allow_unverified_ssl = var.vsphere_allow_insecure
  password             = var.vsphere_password
  user                 = var.vsphere_username
  vsphere_server       = var.vsphere_server
  version = "~> 1.17.4"
}


# Data sources

data "vsphere_datacenter" "dc" {
  name = var.vsphere_datacenter
}

data "vsphere_datastore" "datastore" {
  datacenter_id = data.vsphere_datacenter.dc.id
  name          = var.vsphere_datastore[count.index]
  count         = length(var.vsphere_datastore)
}

data "vsphere_network" "management" {
  datacenter_id = data.vsphere_datacenter.dc.id
  name          = var.management_net_id
}

data "vsphere_resource_pool" "pool" {
  count         = length(var.vsphere_resource_pool_name)
  datacenter_id = data.vsphere_datacenter.dc.id
  name          = var.vsphere_resource_pool_name[count.index]
}

data "vsphere_virtual_machine" "template" {
  datacenter_id = data.vsphere_datacenter.dc.id
  name          = var.image
}

# Resource declarations

resource "vsphere_virtual_machine" "my_vms" {
  clone {
    template_uuid    = data.vsphere_virtual_machine.template.id
  }
  count              = 3
  datastore_id       = data.vsphere_datastore.datastore[count.index].id
  disk {
    datastore_id     = data.vsphere_datastore.datastore[count.index].id
    label            = "disk0"
    keep_on_remove   = false
    size             = var.boot_disk_size
    thin_provisioned = data.vsphere_virtual_machine.template.disks[0].thin_provisioned
  }
  folder             = var.vsphere_folder[count.index]
  guest_id           = data.vsphere_virtual_machine.template.guest_id
  memory             = var.memory
  name               = var.node_names[count.index]
  network_interface {
    network_id       = data.vsphere_network.management.id
  }
  num_cpus           = var.cpu
  resource_pool_id   = data.vsphere_resource_pool.pool[count.index].id
}

```

### Debug Output
See https://gist.github.com/andrew-lee-metaswitch/16751da396946caac12683e0d4a8896d

### Expected Behavior

Applied and created the first VM in a resource

### Actual Behavior

Failed to create the first VM

### Important Factoids

I've been doing some code-reading:

L519 of the gist corresponds to starting running `resourceVSphereVirtualMachineCustomizeDiff` in `resource_vsphere_virtual_machine.go`

This function calls `virtualdevice.DiskDiffOperation`, which is defined in vsphere/internal/virtualdevice/virtual_machine_disk_subresource.go (line 563 of the gist)

This diff is successful L575: ` DiskDiffOperation: Disk diff customization complete, sending new diff`
Then the whole `resourceVSphereVirtualMachineCustomizeDiff` is successful: see L603:
`vsphere_virtual_machine (ID = <new resource>): Diff customization and validation complete`

If I run terraform plan without any target I get:
```
      + disk {
          + attach           = false
          **+ datastore_id     = "datastore-13037"**
          + device_address   = (known after apply)
          + disk_mode        = "persistent"
          + disk_sharing     = "sharingNone"
          + eagerly_scrub    = false
          + io_limit         = -1
          + io_reservation   = 0
          + io_share_count   = 0
          + io_share_level   = "normal"
          + keep_on_remove   = false
          + key              = 0
          + label            = "disk0"
          + path             = (known after apply)
          + size             = 20
          **+ thin_provisioned = true**
          + unit_number      = 0
          + uuid             = (known after apply)
          + write_through    = false
        }

```

If I run terraform apply with a specific target I get, this

```
      + disk {
          + attach           = false
          **+ datastore_id     = "<computed>"**
          + device_address   = (known after apply)
          + disk_mode        = "persistent"
          + disk_sharing     = "sharingNone"
          + eagerly_scrub    = false
          + io_limit         = -1
          + io_reservation   = 0
          + io_share_count   = 0
          + io_share_level   = "normal"
          + keep_on_remove   = false
          + key              = 0
          + label            = "disk0"
          + path             = (known after apply)
          + size             = 20
          **+ thin_provisioned = false**
          + unit_number      = 0
          + uuid             = (known after apply)
          + write_through    = false
        }
```
I then approve the change and then hit the error. 
So why is the graph produced by `GraphTypePlan` different to the graph produced by `GraphTypeApply`?

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by tenthirtyam** [comment]
@andrew-lee-metaswitch - Do you still have the same issue in the latest version of the provider?

**Comment by andrew-lee-metaswitch** [comment]
This is now fixed (with terraform 1.1)  With both the downlevel and the latest plugin I appear to be able to deploy but they both have the following debug log:

2022-03-04T13:11:02.858Z [WARN]  Provider "registry.terraform.io/hashicorp/vsphere" produced an invalid plan for vsphere_virtual_machine.my_vms[0], but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .tools_upgrade_policy: planned value cty.StringVal("manual") for a non-computed attribute
      - .ept_rvi_mode: planned value cty.StringVal("automatic") for a non-computed attribute
      - .scsi_controller_count: planned value cty.NumberIntVal(1) for a non-computed attribute
      - .num_cores_per_socket: planned value cty.NumberIntVal(1) for a non-computed attribute
      - .run_tools_scripts_after_power_on: planned value cty.True for a non-computed attribute
      - .cpu_share_level: planned value cty.StringVal("normal") for a non-computed attribute
      - .memory_limit: planned value cty.NumberIntVal(-1) for a non-computed attribute
      - .scsi_type: planned value cty.StringVal("pvscsi") for a non-computed attribute
      - .cpu_limit: planned value cty.NumberIntVal(-1) for a non-computed attribute
      - .latency_sensitivity: planned value cty.StringVal("normal") for a non-computed attribute
      - .memory_share_level: planned value cty.StringVal("normal") for a non-computed attribute
      - .wait_for_guest_ip_timeout: planned value cty.NumberIntVal(0) for a non-computed attribute
      - .wait_for_guest_net_routable: planned value cty.True for a non-computed attribute
      - .ide_controller_count: planned value cty.NumberIntVal(2) for a non-computed attribute
      - .scsi_bus_sharing: planned value cty.StringVal("noSharing") for a non-computed attribute
      - .force_power_off: planned value cty.True for a non-computed attribute
      - .firmware: planned value cty.StringVal("bios") for a non-computed attribute
      - .swap_placement_policy: planned value cty.StringVal("inherit") for a non-computed attribute
      - .run_tools_scripts_before_guest_shutdown: planned value cty.True for a non-computed attribute
      - .sata_controller_count: planned value cty.NumberIntVal(0) for a non-computed attribute
      - .boot_retry_delay: planned value cty.NumberIntVal(10000) for a non-computed attribute
      - .wait_for_guest_net_timeout: planned value cty.NumberIntVal(5) for a non-computed attribute
      - .run_tools_scripts_before_guest_standby: planned value cty.True for a non-computed attribute
      - .migrate_wait_timeout: planned value cty.NumberIntVal(30) for a non-computed attribute
      - .poweron_timeout: planned value cty.NumberIntVal(300) for a non-computed attribute
      - .run_tools_scripts_after_resume: planned value cty.True for a non-computed attribute
      - .shutdown_wait_timeout: planned value cty.NumberIntVal(3) for a non-computed attribute
      - .hv_mode: planned value cty.StringVal("hvAuto") for a non-computed attribute
      - .clone[0].timeout: planned value cty.NumberIntVal(30) for a non-computed attribute
      - .disk[0].unit_number: planned value cty.NumberIntVal(0) for a non-computed attribute
      - .disk[0].write_through: planned value cty.False for a non-computed attribute
      - .disk[0].attach: planned value cty.False for a non-computed attribute
      - .disk[0].disk_sharing: planned value cty.StringVal("sharingNone") for a non-computed attribute
      - .disk[0].io_share_level: planned value cty.StringVal("normal") for a non-computed attribute
      - .disk[0].io_share_count: planned value cty.NumberIntVal(0) for a non-computed attribute
      - .disk[0].controller_type: planned value cty.StringVal("scsi") for a non-computed attribute
      - .disk[0].eagerly_scrub: planned value cty.False for a non-computed attribute
      - .disk[0].io_limit: planned value cty.NumberIntVal(-1) for a non-computed attribute
      - .disk[0].disk_mode: planned value cty.StringVal("persistent") for a non-computed attribute
      - .disk[0].io_reservation: planned value cty.NumberIntVal(0) for a non-computed attribute
      - .network_interface[0].bandwidth_limit: planned value cty.NumberIntVal(-1) for a non-computed attribute
      - .network_interface[0].bandwidth_share_level: planned value cty.StringVal("normal") for a non-computed attribute
      - .network_interface[0].adapter_type: planned value cty.StringVal("vmxnet3") for a non-computed attribute
      - .network_interface[0].bandwidth_reservation: planned value cty.NumberIntVal(0) for a non-computed attribute

Googling - this seems to be related to https://github.com/hashicorp/terraform/pull/27177 which went into terraform 0.14.0. 

Whatever - this isn't a problem for us.

**Comment by tenthirtyam** [comment]
Hi, @andrew-lee-metaswitch! 👋🏻 

Many thanks for the response and update to the issue, it's appreciated.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.
